taxcalc/simpletaxio.py

```diff
@@ -0,0 +1,648 @@
+        # read input file contents into self._input dictionary
+        self._read_input(input_filename)
+        self._params = Parameters()
+        # implement reform if reform file is specified
+        if reform_filename:
+            reform = SimpleTaxIO.dictionary_from_reform_file(reform_filename)
+            self._params.implement_reform(reform)
+        # validate input variable values
+        self._validate_input()
+        self._calc = self._calc_object()
+        self._output = {}
+
+    def calculate(self, calc_marginal_tax_rates=False, write_output_file=True):
+        """
+        Calculate taxes for all INPUT lines and write OUTPUT to file.
+
+        Parameters
+        ----------
+        calc_marginal_tax_rates: boolean
+
+        write_output_file: boolean
+
+        Returns
+        -------
+        nothing: void
+        """
+        # loop through self._year_set doing tax calculations and saving output
+        for calcyear in self._year_set:
+            if calcyear != self._calc.params.current_year:
+                self._calc.params.set_year(calcyear)
+            self._calc.calc_all()
+            calc_records_tax_year = getattr(self._calc.records, 'FLPDYR')
+            for idx in range(0, self._calc.records.dim):
+                indyr = calc_records_tax_year[idx]
+                if indyr == calcyear:
+                    lnum = idx + 1
+                    ovar = SimpleTaxIO._extract_output(self._calc.records, idx,
+                                                       self._input[lnum])
+                    self._output[lnum] = ovar
+            if calc_marginal_tax_rates:
+                (mtr_fica, mtr_itax, _) = self._calc.mtr('e00200')
+                for idx in range(0, self._calc.records.dim):
+                    indyr = calc_records_tax_year[idx]
+                    if indyr == calcyear:
+                        lnum = idx + 1
+                        self._output[lnum][7] = 100 * mtr_itax[idx]
+                        self._output[lnum][9] = 100 * mtr_fica[idx]
+        # write contents of self._output
+        if write_output_file:
+            self._write_output_file()
+
+    def number_input_lines(self):
+        """
+        Return number of lines read from INPUT file.
+        """
+        return len(self._input)
+
+    @staticmethod
+    def dictionary_from_reform_file(reform_filename):
+        """
+        Read reform file, strip //-comments, and return dict based on JSON.
+        The reform file is JSON with string policy-parameter primary keys and
+           string years as secondary keys.
+        Returned dictionary has integer years as primary keys and
+           string policy-parameters as secondary keys.
+        The returned dictionary is suitable as the argument to the Parameters
+           class implement_reform(reform_dict) method.
+        """
+        # check existence of specified reform file
+        if not os.access(reform_filename, os.F_OK):
```

Comment:
@talumbau,  Good suggestion to move away from os.access(...), but I've decided to use os.path.isfile instead of os.path.exists, which returns true if the supplied path is a directory (not a file).  Same changes are made in several places, so I'll thank you for your suggestion just once here.